Add training options

luckybit4755 opened the pull request:
### Continue training existing model

Modified train_unconditional so the model is reloaded if it exists:

`export COMMAND="python examples/train_unconditional.py --resolution 32 --num_epochs 10 --train_data_dir training-images --output_dir model"`

```
% ${COMMAND}
creating fresh model
....
```

```
% ${COMMAND}
reloading model from model/unet
reloaded model from model/unet
....
```
### Checkpoint periodically on model save

Checkpoint a copy of the model on save:
% ${COMMAND} --checkpoint_model_epochs 10
...
checkpointed model/unet to model/checkpoints/checkpoint-2022-08-18+17-17-48
...

The arguments should be n * save_model_epochs ; where n is an integer >0 or 0 to disable (default)

```
% tree model/checkpoints
model/checkpoints
└── checkpoint-2022-08-18+17-17-48
    ├── config.json
    └── diffusion_pytorch_model.bin

1 directory, 2 files
```

### Timestamp test_samples

The test_samples are also timestamped to allow visual inspection over time:

`% ${COMMAND} --timestamp_test_samples`

Show up with names like:

-  test_samples-2022-08-18+17-49-52+000000
-  test_samples-2022-08-18+17-55-34+000009

### Script for generating images:

```
% ./scripts/generate_images.py model 3
modification time on model is 2022-08-18+17-17-48
loading the model from model
loaded the model from model
creating image and saving to generated/model/2022-08-18+17-17-48/image-0000.png
100%|####| 1000/1000 [00:16<00:00, 60.02it/s]
image saved to generated/model/2022-08-18+17-17-48/image-0000.png
creating image and saving to generated/model/2022-08-18+17-17-48/image-0001.png
100%|####| 1000/1000 [00:17<00:00, 65.98it/s]
image saved to generated/model/2022-08-18+17-17-48/image-0001.png
creating image and saving to generated/model/2022-08-18+17-17-48/image-0002.png
100%|####| 1000/1000 [00:17<00:00, 65.31it/s]
image saved to generated/model/2022-08-18+17-17-48/image-0002.png
writing html to generated/model/2022-08-18+17-17-48/images.html
wrote html to generated/model/2022-08-18+17-17-48/images.html
```

The directory name is based off of the model and timestamp of it's directory.

Successive runs will not clobber existing files but skip over them.

patrickvonplaten commented:
@anton-l could you check here?

anton-l commented on scripts/generate_images.py:
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python
```
Since every model's pipeline is quite specific, we probably won't be adding inference scripts to keep things more maintainable. Instead we're relying on example snippets on the main README and the (still WIP) documentation. 
If that makes sense, would you mind removing this file? :slightly_smiling_face:

luckybit4755 replied:
Sure thing, I have a side directory where I keep other scripts. Seeing generate_logits.py in scripts was the motivation to add it there

anton-l replied:
Ah, I see! Yeah, we should remove generate_logits too, it's a rogue debugging script :smile:

anton-l commented on examples/train_unconditional.py:
```diff
@@ -171,15 +181,26 @@ def transforms(examples):
 
                 # denormalize the images and save to tensorboard
                 images_processed = (images * 255).round().astype("uint8")
+                epoch06 = str(epoch).zfill( 6 )
+                if args.timestamp_test_samples:
+                    test_samples = datetime.now().strftime(f"test_samples-%Y-%m-%d+%H-%M-%S+{epoch06}")
```
This looks like a bit of leftover debugging code (the `06` part in particular`)?

luckybit4755 replied:
The 06 bit is to keep the test_samples names so that they can be sorted the same either alphabetically or numerically. It should be under the if statement for timestamp_test_samples however.

anton-l commented on examples/train_unconditional.py:
```diff
@@ -34,29 +36,37 @@ def main(args):
+
+    model_dir = f"{args.output_dir}/unet"
+    if os.path.exists(model_dir):
+        print(f"reloading model from {model_dir}")
+        model = UNet2DModel.from_pretrained(model_dir)
+        print(f"reloaded model from {model_dir}")
```
Instead of print statements it's better to use `logger.info(...)` to make the amount of logging controllable by the user.

luckybit4755 commented:
I have not worked with transformers before and implemented checkpoints a little differently than I see in [run_glue_no_trainer.py](https://github.com/huggingface/transformers/blob/6667b0d7bf0a135b9a85d2d21b7eb25ec8ad58cb/examples/pytorch/text-classification/run_glue_no_trainer.py#L590-L594) and implemented it like [so](https://github.com/luckybit4755/diffusers/blob/a452dc9a1b9b280da26daffd8bf51c8d176c1083/examples/train_unconditional.py#L198-L203) 

In the interests of consistency it might make sense to redo the entire checkpoint / reload mechanism along the lines of run_glue_no_trainer 

I'll happily mirror that logic if it will save you time so you can work on other things, but if you want to implement it yourself so it completely matches the project gestalt

anton-l commented:
Hey @luckybit4755 sorry for the late reply! Yeah, feel free to copy the logic over, or I can include the changes into your PR if you don't mind :)

luckybit4755 commented:
Yeah, that's totally fine if you wouldn't mind. Sorry been a bit distracted as well.

patrickvonplaten commented:
Do we want to copy the "resume" training logic here or leave for now? cc @anton-l
